File: internal/repository/category.go
package repository

import (
	"context"

	"things-expired/internal/model"
	"gorm.io/gorm"
)

// ICategoryRepository 分类仓储接口
type ICategoryRepository interface {
	Create(ctx context.Context, category *model.Category) error
	GetByID(ctx context.Context, id uint) (*model.Category, error)
	GetByUserIDAndName(ctx context.Context, userID uint, name string) (*model.Category, error)
	List(ctx context.Context, userID uint, page, pageSize int) ([]*model.Category, int64, error)
	Update(ctx context.Context, category *model.Category) error
	Delete(ctx context.Context, id uint) error
}

// CategoryRepository 分类仓储实现
type CategoryRepository struct {
	db *gorm.DB
}

// NewCategoryRepository 创建分类仓储
func NewCategoryRepository(db *gorm.DB) ICategoryRepository {
	return &CategoryRepository{db: db}
}

func (r *CategoryRepository) Create(ctx context.Context, category *model.Category) error {
	return r.db.WithContext(ctx).Create(category).Error
}

func (r *CategoryRepository) GetByID(ctx context.Context, id uint) (*model.Category, error) {
	var category model.Category
	if err := r.db.WithContext(ctx).First(&category, id).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, nil
		}
		return nil, err
	}
	return &category, nil
}

func (r *CategoryRepository) GetByUserIDAndName(ctx context.Context, userID uint, name string) (*model.Category, error) {
	var category model.Category
	if err := r.db.WithContext(ctx).Where("user_id = ? AND name = ?", userID, name).First(&category).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, nil
		}
		return nil, err
	}
	return &category, nil
}

func (r *CategoryRepository) List(ctx context.Context, userID uint, page, pageSize int) ([]*model.Category, int64, error) {
	var categories []*model.Category
	var total int64

	query := r.db.WithContext(ctx).Model(&model.Category{}).Where("user_id = ?", userID)

	if err := query.Count(&total).Error; err != nil {
		return nil, 0, err
	}

	offset := (page - 1) * pageSize
	if err := query.Offset(offset).Limit(pageSize).Order("sort_order ASC, created_at DESC").Find(&categories).Error; err != nil {
		return nil, 0, err
	}

	return categories, total, nil
}

func (r *CategoryRepository) Update(ctx context.Context, category *model.Category) error {
	return r.db.WithContext(ctx).Save(category).Error
}

func (r *CategoryRepository) Delete(ctx context.Context, id uint) error {
	return r.db.WithContext(ctx).Delete(&model.Category{}, id).Error
}

File: src/repository/category.ts
import { eq, and, like, sql } from "drizzle-orm";
import type { DBInstance } from "@/repository/db";
import { categories } from "@/model/category";
import type { ICategoryRepository } from "@/repository/interfaces";
import type { InsertCategory, Category } from "@/model/category";

export class CategoryRepository implements ICategoryRepository {
  constructor(private db: DBInstance) {}

  async findById(id: number): Promise<Category | null> {
    const result = await this.db
      .select()
      .from(categories)
      .where(eq(categories.id, id))
      .limit(1);
    return result[0] || null;
  }

  async findByUserIdAndName(
    userId: number,
    name: string,
  ): Promise<Category | null> {
    const result = await this.db
      .select()
      .from(categories)
      .where(
        and(eq(categories.userId, userId), eq(categories.name, name)),
      )
      .limit(1);
    return result[0] || null;
  }

  async create(data: InsertCategory): Promise<Category> {
    const result = await this.db
      .insert(categories)
      .values(data)
      .returning();
    return result[0] as Category;
  }

  async update(
    id: number,
    data: Partial<InsertCategory>,
  ): Promise<Category | null> {
    const result = await this.db
      .update(categories)
      .set(data)
      .where(eq(categories.id, id))
      .returning();
    return result[0] || null;
  }

  async delete(id: number): Promise<void> {
    await this.db.delete(categories).where(eq(categories.id, id)).run();
  }

  async list(params: {
    userId: number;
    page: number;
    pageSize: number;
    keyword?: string;
  }): Promise<{ items: Category[]; total: number }> {
    const { userId, page, pageSize, keyword } = params;
    const offset = (page - 1) * pageSize;

    const conditions = [eq(categories.userId, userId)];
    if (keyword) {
      conditions.push(like(categories.name, `%${keyword}%`));
    }

    const items = await this.db
      .select()
      .from(categories)
      .where(and(...conditions))
      .limit(pageSize)
      .offset(offset)
      .orderBy(categories.sortOrder);

    const countResult = await this.db
      .select({ count: sql<number>`count(*)` })
      .from(categories)
      .where(and(...conditions));

    const count = countResult[0]?.count ?? 0;

    return { items, total: count };
  }
}
